### PR TITLE
Reproduceable testcase that is crashing Aiken JNA Wrapper.

### PR DESCRIPTION
**Pull request by matiwinnetou**
1. The compiled transaction is using Aiken SHA1: bb820ebdd8636ee33153d27b93ed6cd63149a039
2.  All parameters are main-net params including initial budget and minFeeA and minFeeB
3. It was running create batches from votes from hydra-voting-example, where batch size: 3
4. Contract compiled with aiken build -k, meaning all trace statements are compiled in

Hunch:
My hunch is that error is actually caused by pallas code together with the fact that we are using aiken build -k, so compiled in traces.

Questions:
1. Why do we actually need to use pallas, is this necessary, pallas could / is buggy and is extra dependency (sorry I didn't dig deeper into this). Afaik we only need ExUnits so this is minimal interface we need from this library. Bloxbean after all is only returning us ExUnits, not the whole serialised redeemers in hex or?

Aiken's compiled contract:
```
59115201000032323232323232323232323232323232222533300a32323232323232323232323232323232323232323232323232323253330253370e9000000899191919180999817999814a99981499980600e80e00f0a51153302b49014b726573756c745f73656e745f746f5f6f776e5f616464726573735f6d6174636828696e707574732c206f7574707574732c206f75747075745f7265666572656e636529203f2046616c73650014a0980103d87a80004c0103d87980003302f3330295333029323232323232323232323232323322323253303c3372c9210e6f6e5f636861696e5f686173683a00300537520022a6607866e592410f6f66665f636861696e5f686173683a0030053752006266e3c00400cdd71981a981b800a40046602a004460326ecc00403c0408c8dcc80099801000a441003001001222533333303f00213232323232323300e002001337149101012800002533303a337100069007099b80483c80400c54ccc0e8cdc4001a410004266e04cdc0241002800690070b299981f000899b8a4881035b5d2900004132323233714911015b00001375c607e0046eb4c0f4004ccc03c00488cdd81ba848160dd499980400400119b8b375a607c0026eb8c0fc004cdd81ba8482e804dd499b8a4881012900004375860780046eb4c0e8004c8cdd81ba8303a001374e60760026ea80084c94ccc0f00044cdc52441027b7d00002132323233714911027b2000001375c607a0046eb8c0ec004ccc01c00488c8cdd81ba94881022c2000375266600e00e607a00666e29221023a2000001333006006303d002337146eb8c0f0004dd7181e80099bb03752910102207d0037520046eac0084c94ccc0f00044cdc52441025b5d00002132323233714911015b00001375c607a0046eb8c0ec004ccc01400488cdd81ba94881022c2000375266600c00c00466e28dd7181e0009bae303d001337606ea5221015d0037520046eb00084c8cc024004008dd68010991919b8a48810268270000133300b001337026e34005200233714911012700002375c00460020024444a66607400620022646466008004666600c00c002008006607c008607800660020024444a66607000620022646466008004666600c00c002008006607800860740066002002444a66606066e1c0092000153330303370e6e340052000148810130001001133300300333708004900a19b8b3370066e14009201448180004c00400488894ccc0bccdc4001240002002264646464666601001000e66e0401920020013371666e000094ccc0c8cdc400124028290300a40dc66e2ccdc0000a99981919b8800148050520601481b8010cdc30012404066e0c00520203371c00600460020024444a66606400620022646466008004666600c00c002008006606c0086068006294454cc0ad2412976616c69646174655f766f74657328766f7465732c2062617463685f6861736829203f2046616c73650014a0980103d87a80004c0103d87980004bd7019191919191919191919801000a5eb7bdb180024c0040048894ccc0d800840044c8c8c8c8c8cccc088ccc020020010018cccc8888cccc03401000c0080052f5bded8c066e9520003303c3750004660786ea00052f5c066e9520003303c37506eb4cc0c8c0d000d20023303c37506eb4cc0c8c0d000d20043303c37506eb4cc0c8c0d000d20064bd700140138141bad33031303300448018dd6998181819001a4008646464646466e9520003303e0043303e37500066607c6ea0008cc0f8dd4000a5eb80c8cc03400520003300d00448000c8cc03000520003300c00348008c8cc02c00520003300b00248010cdd2a4000660746ea0dd6998181819000a4008660746ea0dd6998181819000a400c97ae0002303a003303800230010012222253330360041330373376000600497adef6c6013232323253330353375e6600a00e00266e9520004bd7009981d99bb0007006008153330353375e00e00226607666ec001c01800c4cc0eccdd800080119999804804801803803002981b801981b801181d002981c0021119299981719b874800800440084c8c8004dd6981b000981600198160011119199119299981819b87480080044cdd2a400497ae0132323374a90001981c1ba830040014bd701bad3038001302e003302e0020012375a660546058008900119801a5eb84101000081010100810102001180b9981999981699b87375a6605260560069004000a60103d87a80004c0103d87980003303333302d3370e004002980103d87a80004c0103d87980004bd7018008009112999818001099ba5480092f5c026464a66605a6006004266e95200033033375000497ae013330050050010033034003375a606400464646466004002464646605c64a66605c66e1d2000001132323232323232323232533303d3040002149854cc0e92401334c6973742f5475706c652f436f6e73747220636f6e7461696e73206d6f7265206974656d73207468616e2065787065637465640016375a607c002607c0046eb4c0f0004c0f0008dd6981d000981d0011bad30380013038002375c606c00260580042a6606092012b436f6e73747220696e64657820646964206e6f74206d6174636820616e7920747970652076617269616e740016302c0010013033001302932533302c3370e9002181580088008a998172492a4578706563746564206f6e20696e636f727265637420636f6e7374727563746f722076617269616e742e001633027302933027302900148009200401d3001001222533302f00214bd7009919198191801801199802802800801981980198188011bae302d001302301d153330253370e9001000899191919191919180b19819199816299981619980781000f8108a51153302e49014b726573756c745f73656e745f746f5f6f776e5f616464726573735f6d6174636828696e707574732c206f7574707574732c206f75747075745f7265666572656e636529203f2046616c73650014a0980103d87a80004c0103d87980003303233302c533302c332232323371e0060026eb8cc0acc0b400520023300b0022300f376600200400c294454cc0b924013176616c69646174655f726573756c747328616c6c5f726573756c74732c2062617463685f6861736829203f2046616c73650014a0980103d87a80004c0103d87980004bd7019191991191998020019119ba548000cc0e0dd3199980f1bab3302e303000248000dd5998171818000a4000046048660706ea52210033038375000897ae00013374a90001981a9ba64bd6f7b6301981a9ba94890033035375000297ae000300430010012222533303400310011323233330060060010043300400200330380043036003323233223300300223232330323253330323370e90000008991919191919299981e982000109981c99819002919191981e19299981e19b87480000044c8c8c8c94ccc114c1200085261533042491334c6973742f5475706c652f436f6e73747220636f6e7461696e73206d6f7265206974656d73207468616e2065787065637465640016375a608c002608c0046eb4c110004c0e800c54cc0f92412b436f6e73747220696e64657820646964206e6f74206d6174636820616e7920747970652076617269616e740016303a0023303c32533303c3370e90000008991919191919299982398250010a4c2a66088921334c6973742f5475706c652f436f6e73747220636f6e7461696e73206d6f7265206974656d73207468616e2065787065637465640016375a609000260900046eb4c118004c118008dd69822000981d0010a9981f24812b436f6e73747220696e64657820646964206e6f74206d6174636820616e7920747970652076617269616e740016303a001498c0f8008c0f000526153303a491334c6973742f5475706c652f436f6e73747220636f6e7461696e73206d6f7265206974656d73207468616e2065787065637465640016375a607c002607c0046eb8c0f0004c0f0008dd5981d00098180010a9981a24812b436f6e73747220696e64657820646964206e6f74206d6174636820616e7920747970652076617269616e740016303000153330313370e6eb4cc0b4c0bc005200400413374a90001981b800a5eb804cdd2a400497ae03037001302d3253330303370e9002181780088008a9981924812a4578706563746564206f6e20696e636f727265637420636f6e7374727563746f722076617269616e742e00163302b302d3302b302d0014800920040200043001001222533303200214bd70099191919299981899b87480080044ccc01c01c00c0144c8c8cc0e4004ccc02402401401cc0e4004c0bc008c0bc004c00c008c0d800cc0d0008cdc0000a40046eb4c0bc004c0bc008dd71816800981180e8a50302301c223232323374a9000198171ba70033302e375200297ae0333222333008003322233003001002002001002223322300b332233714004002004002004002002600c910100330060022300200130010012222533302a003100113232333300600600100433004002003302e004375c605800646e48004c0040048894ccc09800852f5c026464660526ea4c00c008ccc01401400400cc0a800cc0a0008888c8c8cc0180048c8c8cdc78008029bae302b00130213253330243370e9001181180088008a9981324812a4578706563746564206f6e20696e636f727265637420636f6e7374727563746f722076617269616e742e00163301f30213301f3021001480012000330070032323253330243370e9001000899251302200214a060440026603c60406603c60400029000240006644646464640026eb8c0b0004c088c94ccc094cdc3a4004604800220022a6604e92012a4578706563746564206f6e20696e636f727265637420636f6e7374727563746f722076617269616e742e001633020302233020302233020302200148009200048000c0a8004c080c94ccc08ccdc3a4000604400220022a6604a9212a4578706563746564206f6e20696e636f727265637420636f6e7374727563746f722076617269616e742e00163300a00223375e6603e60420029000001001800980080091129998118010a50132325333020300300214a2266600a00a002006604e006604a0046002002444a666042004297ae013232533301e30030021330240023330050050010031333005005001003302500330230023001001222533301f00213374a900125eb804c8c94ccc070c00c0084cdd2a40006604400497ae0133300500500100330230033021002300100122533301c00114a226464a6660320042660080080022940c080008cdc3a400460346ea8c0780048888cccc01401000c008004c004004888894ccc070010400c4c8c8c8c8ccccc02402400cccccc02801c004008018014018014c07400cc074008c080014c078010c0040048888894ccc06c0144cc070cdd8002001a5eb7bdb1804c8c8c8c94ccc068cdd79980280400099ba5480012f5c026604066ec002001c02454ccc068cdd780400089919299980e19b87480000044c8c8cc090cdd80060008039812000980d0010802980d00099980300400380109981019bb000100233333300a00a003008007006005301c003301c002301f006301d0052223374a90001980c19ba548000cc060dd419b80375a6601c602000490001bad3300e301000148000cc060dd419b80375a6601c602000490011bad3300e301000148008cc060dd419b80375a6601c602000490021bad3300e3010001480112f5c097ae0223322533301233720004002266e9520004bd700a99980919b8f00200113374a900125eb804cdd2a400897ae037660046ecc004dd61980498059980498058022400090021bac33008300a33008300a0034800120003012001300832533300b3370e9001180500088008a99806a492a4578706563746564206f6e20696e636f727265637420636f6e7374727563746f722076617269616e742e001633006300800148008526163001001222533300d00214984cc024c004c03c008ccc00c00cc040008004cc0040052000222233330073370e00200601c4666600a00a66e00011200230100010020022300737540024600a6ea80055cd2b9b5738aae7555cf2ab9f5740ae855d12ba41
```

### CODE REVIEW

**Comment by satran004** [comment]
> 

If I am not wrong Aiken is using pallas internally, so we need to parse all hex inputs to required object and pass to aiken code.

**Comment by satran004** [comment]
> We should have our Err wrapper and use UPLC error as a delegate I think:
> 
> image

We can do that. But finally we need an error string to pass it back to Java layer.

**Comment by matiwinnetou** [comment]
@satran004 Does JNI allows us to throw exceptions from Rust to java?

File: src/lib.rs [hunk 1]
```diff
@@ -22,8 +22,8 @@ pub struct SlotConfig {
 #[allow(missing_copy_implementations)]
 #[derive(Clone)]
 pub struct InitialBudget {
-    pub mem: u64,
```
Comment (matiwinnetou) [inline src/lib.rs:25]: @satran004 I can isolate this in smaller PR or we leave it for now?

**Comment by satran004** [comment]
> @satran004 Does JNI allows us to throw exceptions from Rust to java?

AFAIK, native exception has to be converted to Java exception in JNI layer.

But we are using JNA here. Check for "Obtaining "last" error" section in this link 
http://java-native-access.github.io/jna/5.13.0/javadoc/overview-summary.html#crash-protection
I have not tried this. 

But I think for our case, the simple solution is to serialize the error to a string and print it in Java side as we can't do much in Java in case of native unknown error.

File: src/lib.rs (hunk 1 above)
Comment (satran004) [reply src/lib.rs:25]: @matiwinnetou Yes. I think we can keep this change in a separate PR.

**Pull request closed without merging**